### PR TITLE
Don't log secret data

### LINKED ISSUE #14: Don't log secret data
https://github.com/presslabs/controller-util/blob/83fa29ebb62b339a5ad94feb43ec7714da15321b/syncer/object.go#L55

### PR DESCRIPTION
**Pull request by rstefan1**
fixes #14

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -26,6 +26,7 @@ linters:
     - goimports
     - maligned
     - gochecknoglobals
+    - godox
```
Comment (bogdanpetrea) [inline .golangci.yml]: This should be in a different PR.

File: syncer/syncer.go [hunk 2]
```diff
@@ -26,6 +26,10 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/client-go/tools/record"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	// TODO: Controller-runtime/log si deprecated. We should use pkg/log/
```
Comment (bogdanpetrea) [inline syncer/syncer.go]:
```suggestion
	// TODO: Controller-runtime/log is deprecated. We should use pkg/log/
```

File: syncer/object.go [hunk 3]
```diff
@@ -53,19 +68,19 @@ func (s *ObjectSyncer) Sync(ctx context.Context) (SyncResult, error) {
...
 		err = nil
 	} else if err == ErrIgnore {
-		log.V(1).Info("syncer skipped", "key", key, "kind", fmt.Sprintf("%T", s.Obj))
+		log.V(1).Info("syncer skipped", "key", key, "kind", fmt.Sprintf("%T", objectWithoutSecretData(s.Obj)))
 		err = nil
 	} else if err != nil {
 		result.SetEventData(eventWarning, basicEventReason(s.Name, err),
 			fmt.Sprintf("%T %s failed syncing: %s", s.Obj, key, err))
-		log.Error(err, string(result.Operation), "key", key, "kind", fmt.Sprintf("%T", s.Obj), "diff", diff)
+		log.Error(err, string(result.Operation), "key", key, "kind", fmt.Sprintf("%T", objectWithoutSecretData(s.Obj)), "diff", diff)
```
Comment (vtemian) [inline syncer/object.go]: It's a good approach, but maybe we can do it somehow more generic.

The idea would be to reduce the leaking of secrets in logging, thus meaning that the entire obfuscation should take place at the logging component level. Only by doing it in this way we can be 100% sure that we don't leak secrets.

Since you can't hijack the logging in Go and implementing a logging component introduces too much overhead and complexity, we can keep it this way.

Just some small improvements:
  - `ToString` instead of `objectWithoutSecretData`
  - `ToString` should return the `fmt.Sprintf("%T")` of the object. If it fails, return empty string.

Of course, we'll need some tests for it, but unittests will do the trick.

File: syncer/object.go [hunk 4]
```diff
@@ -30,11 +31,44 @@ type ObjectSyncer struct {
 	previousObject runtime.Object
 }
 
+// objectWithoutSecretData returns the object without secretData
+func objectWithoutSecretData(obj runtime.Object) runtime.Object {
```
Comment (vtemian) [inline syncer/object.go]: `objectWithoutSecretData` implies more creating an object without secret data.
I would rather call it `stripSecrets` or something similar.

File: syncer/object.go [hunk 5]
```diff
@@ -30,11 +31,44 @@ type ObjectSyncer struct {
...
+	// if obj is secret, don't print secret data
```
Comment (vtemian) [inline syncer/object.go]: This function doesn't print any data.

File: syncer/object.go [hunk 6]
```diff
@@ -30,11 +31,44 @@ type ObjectSyncer struct {
...
+
+		return s
+	}
+
+	return obj
+}
+
 // GetObject returns the ObjectSyncer subject
-func (s *ObjectSyncer) GetObject() interface{} { return s.Obj }
+func (s *ObjectSyncer) GetObject() interface{} {
```
Comment (vtemian) [inline syncer/object.go]: ```but it's neither idiomatic nor necessary to put Get into the getter's name. If you have a field called owner (lower case, unexported), the getter method should be called Owner (upper case, exported), not GetOwner```
https://golang.org/doc/effective_go.html

Comment (vtemian) [reply syncer/object.go]: Just `Object` should be enough.

File: syncer/object.go [hunk 7]
```diff
@@ -30,11 +31,44 @@ type ObjectSyncer struct {
...
+	return objectWithoutSecretData(s.Obj)
+}
+
+// GetPreviousObject returns the ObjectSyncer previous subject
+func (s *ObjectSyncer) GetPreviousObject() interface{} {
```
Comment (vtemian) [inline syncer/object.go]: `Previous`
See above.

File: syncer/object.go [hunk 8]
```diff
@@ -30,11 +31,44 @@ type ObjectSyncer struct {
...
+	return objectWithoutSecretData(s.previousObject)
+}
+
+// GetObjectType returns the type of the ObjectSyncer subject
+func (s *ObjectSyncer) GetObjectType() string {
```
Comment (vtemian) [inline syncer/object.go]: `Type`
See above.

File: syncer/object.go [hunk 9]
```diff
@@ -30,11 +31,44 @@ type ObjectSyncer struct {
...
+	return fmt.Sprintf("%T", s.Obj)
+}
 
 // GetOwner returns the ObjectSyncer owner
-func (s *ObjectSyncer) GetOwner() runtime.Object { return s.Owner }
+func (s *ObjectSyncer) GetOwner() runtime.Object {
```
Comment (vtemian) [inline syncer/object.go]: `Owner`
See above.

File: syncer/object.go [hunk 10]
```diff
@@ -30,11 +31,44 @@ type ObjectSyncer struct {
...
+	return s.Owner
+}
+
+// GetOwnerType returns the type of the ObjectSyncer owner
+func (s *ObjectSyncer) GetOwnerType() string {
```
Comment (vtemian) [inline syncer/object.go]: `OwnerType`
See above.
